src/hypercall.rs

```diff
@@ -259,28 +260,49 @@ pub fn copy_argv(path: &OsStr, argv: &[String], syscmdval: &CmdvalParams, mem: &
 }
 
 /// Copies the environment variables into the VM's memory to the destinations specified in `syscmdval`.
-pub fn copy_env(syscmdval: &CmdvalParams, mem: &MmapMemory) {
-	let env_len = std::env::vars_os().count();
+pub fn copy_env(env: &EnvVars, syscmdval: &CmdvalParams, mem: &MmapMemory) {
 	let envp = mem
 		.host_address(syscmdval.envp)
 		.expect("Systemcall parameters for Cmdval are invalid") as *const GuestPhysAddr;
-	let env_addrs = unsafe { std::slice::from_raw_parts(envp, env_len) };
 
-	// Copy the environment variables into the vm memory
-	for (counter, (key, value)) in std::env::vars_os().enumerate() {
-		if counter >= MAX_ARGC_ENVC.try_into().unwrap() {
-			warn!("Environment is larger than the maximum that can be copied to the VM. Remaining environment is ignored");
-			break;
-		}
+	let env_len = match env {
+		EnvVars::Host => std::env::vars_os().count(),
+		EnvVars::Set(map) => map.len(),
+	};
```

Comment:
I think creating a `Vec<(String, String)>` once would be better than consuming the `vars_os` twice. That would also deduplicate the code below.

Reply:
Either way is ugly, but I've followed your suggestion :-)